core/src/main/java/io/crate/types/DataType.java

```diff
@@ -43,6 +79,20 @@
 
     public abstract int compareValueTo(T val1, T val2);
 
+    @Nullable
+    public DataType innerType() {
```

Comment:
Why do we need a `innerType` on the base DataType class?

Reply:
To have a consistent way to solve the inner type of a DataType. Otherwise we need `instanceof`.

Reply:
I'd be in favor of the `instanceof` instead of adding a property that doesn't make sense for the majority of the types. E.g. if we add a `RowType` or extend the ObjectType to know it's inner types, this will be even more confusing because it would have many inner types, not only one.